Improve the signal handling to avoid crashes during stop

pereyra-m opened the pull request:
|Related issue|
|---|
|Closes #26536|

<!--
This template reflects sections that must be included in new Pull requests.
Contributions from the community are really appreciated. If this is the case, please add the
"contribution" to properly track the Pull Request.

Please fill the table above. Feel free to extend it at your convenience.
-->

## Description

To avoid the crash of the `wazuh-modulesd` during the shutdown, some improvements were added:
- The signals are ignored at startup and until the modules have started. This avoids the termination of the module during the `sleep()` calls.
- There is a wait time of 1 second after the modules have started to allow them settle.
- The signal handling setup was moved after the added wait time to avoid calling a stop before the start has finished.
- The signal handling now has a mutex, it will only process one signal and will ignore the rest

**Update**

The original proposal described above might have many changes. A simplified one is under analysis, where the sleep occurs at the stop method and no sig-ignore is configured at startup. 

## Tests

<!--
Depending on the affected components by this PR, the following checks should be selected and marked.
-->

The reproduction steps now don't cause crashes.

![2024-11-01_11-37](https://github.com/user-attachments/assets/5748c5e7-4600-4521-90ca-9422a700e39a)


<!-- Minimum checks required -->
- Compilation without warnings in every supported platform
  - [x] Linux

sebasfalcone commented on src/wazuh_modules/main.c:
```diff
@@ -16,13 +16,21 @@ static void wm_help();                  // Print help.
 int main(int argc, char **argv)
 {
+    // Ignore signals until the modules have started
+    wm_signals_ignore();
```
Note: Performed some testing over this

If the module initialization takes too long or it hungs we are still able to kill the process, I don't know if this desirable or not but its working as intended

pereyra-m replied:
I'm open to comments here. Ignoring the signals also avoids the `Termination` of the process when it receives a signal during the starting phase, but it might change the current behavior too much.

The real fix is in the mutex, I can revert this.

sebasfalcone reviewed (commented):
Amazing work @pereyra-m

vikman90 reviewed (commented):
The proposed solution is on the right track! However, after reviewing the trace, I would suggest considering some alternative approaches.

One concern from the trace is that the `wm_handler` exit routine is executed by the Vuln Detector thread, while ideally, it should be handled by the main thread.

Additionally, the use of `sleep(1)` raises some concerns. This approach could lead to inconsistent behavior across different systems—on some, it might be too long, while on others, it could be insufficient. It appears that a sleep function is being used to mitigate a race condition, which is not an ideal solution.

Given that the modules seem to have a flag for handling graceful shutdowns, I would recommend exploring the replacement of `wm_signals_ignore` with a per-thread signal mask, ensuring proper signal handling across different threads.

Cheers!

pereyra-m commented on src/wazuh_modules/main.c:
```diff
@@ -16,13 +16,21 @@ static void wm_help();                  // Print help.
 
 static int flag_foreground = 0;         // Running in foreground.
 
+static bool wm_sig_received = false;   // Flag to indicate if a signal is being handled.
```
You are right.

pereyra-m commented:
Hi @vikman90 

Thank you for the review.

> One concern from the trace is that the wm_handler exit routine is executed by the Vuln Detector thread, while ideally, it should be handled by the main thread.

I don't understand why this happens. Can you share the backtrace? I think that setting the signal handling in the main thread shouldn't have this behavior. 

> Additionally, the use of sleep(1) raises some concerns. This approach could lead to inconsistent behavior across different systems—on some, it might be too long, while on others, it could be insufficient. It appears that a sleep function is being used to mitigate a race condition, which is not an ideal solution.

I don't like it either, but I thought that if we had some sleeps at the start of other modules, it wouldn't be that bad. I agree that a real coordination of all the modules is much better, because we could know the exact moment when all of them are ready. But this issue isn't reproduced naturally in the product, so I preferred to make fewer changes and not to have a major rework. 
I proposed it here to allow all the modules to settle, but it isn't the most important part of the fix and I can remove it.

> I would recommend exploring the replacement of wm_signals_ignore with a per-thread signal mask, ensuring proper signal handling across different threads.

I proposed the `wm_signals_ignore()` method to avoid a premature shutdown of the module. I meant to ignore all signals until the daemon is ready to process them. I can remove it also, because in my tests, the most important part of the fix is the mutex at `wm_handler()`

pereyra-m commented:
@sebasfalcone @vikman90 

I've uploaded another proposal for your consideration.

Given that making each module start routine notify the main thread that is ready is a big change, I moved the sleep to the stop section. I try to cause the less impact possible here, but without this, the SIGTERM signal happens too fast and the module crashes. Now I don't affect the time it takes the module to start.

Also, I've removed the signal ignore method at startup to have fewer changes in the PR. Consider that sometimes, the process is terminated because the signal is received during a pre-existent sleep and it isn't ignored anymore; but this isn't a crash.

sebasfalcone commented on src/wazuh_modules/main.c:
```diff
@@ -192,15 +183,48 @@ void wm_cleanup()
+    }
+    w_mutex_unlock(&sig_lock);
+
+    // We wait in case the modules haven't completed the initialization
+    sleep(1);
```
We know this is not the best approach, but is the best we can do without inquiring about a big refactor regarding the initialization and stopping process of the modules

The stop-and-start interfaces we currently have are not prepared for this kind of behaviour

sebasfalcone reviewed (approved):
LGTM

The segfault prevented with this approach, in a normal situation it shouldn't happen anymore
